core/src/com/unciv/logic/battle/Battle.kt

```diff
@@ -402,4 +411,59 @@ object Battle {
+        // according to some strategy guide the slinger's withdraw ability is inherited on upgrade,
+        // according to the Ironclad entry of the wiki the Caravel's is lost on upgrade.
+        // therefore: Implement the flag as unique for the Caravel and Destroyer, as promotion for the Slinger.
+        // I want base chance for Slingers to be 133% (so they still get 76% against the Brute)
+        // but I want base chance for navals to be 50% (assuming their attacker will often be the same baseunit)
+        // the diverging base chance is coded into the effect string as (133%) for now, with 50% as default
+        if (attacker !is MapUnitCombatant) return false         // probably redundant assert-like checks
+        if (defender !is MapUnitCombatant) return false
+        if (!attacker.isMelee()) return false
+        if (defender.unit.isEmbarked()) return false            // Embarked units should never use the ability
+        val attackBaseUnit = attacker.unit.baseUnit
+        val defendBaseUnit = defender.unit.baseUnit
+        val withdrawMatch = Regex("""\((\d+)%\)""").find(withdraw)
+        val percentChance = (
+                (if(withdrawMatch!=null) withdrawMatch.groups[1]!!.value.toFloat() else 50f)
+                * defendBaseUnit.strength / attackBaseUnit.strength
+                * defendBaseUnit.movement / attackBaseUnit.movement).toInt()
+        val dice = Random().nextInt(100)
+        if (dice > percentChance) return false                  // Random element now played out
+        val fromTile = defender.getTile()
+        val attTile = attacker.getTile()
+        assert(fromTile in attTile.neighbors)                   // function should never be called with attacker not adjacent to defender
+        // the following yields almost always exactly three tiles in a half-moon shape (exception: edge of map)
+        val candidateTiles = fromTile.neighbors.filterNot { it == attTile || it in attTile.neighbors }
+        if (candidateTiles.isEmpty()) return false              // impossible on our map shapes? No - corner of a rectangular map
+        val toTile = candidateTiles.random()
+        if (!defender.unit.movement.canMoveTo(toTile)) {        // forbid impassable or blocked
+            val blocker = toTile.militaryUnit
+            if (blocker != null) {
+                val notificationString = "[" + defendBaseUnit.name + "] could not withdraw from a [" +
+                        attackBaseUnit.name + "] - blocked by [" +
+                        blocker.owner + "] [" + blocker.name + "]"
+                defender.getCivInfo().addNotification(notificationString, toTile.position, Color.RED)
+            }
+            return false
+        }
+        if (defendBaseUnit.unitType.isLandUnit() && !toTile.isLand) return false        // forbid retreat from land to sea - embarked already excluded
+        if (toTile.isCityCenter()) return false                                         // forbid retreat into city
```

Comment:
Does it make sense to include these checks into `candidateTiles = fromTile.neighbors.filterNot` ?
Otherwise, it decreases chances for the retreat.

Reply:
As indended. Should be commented in code and #2203. Docs I can find say that chances to withdraw go down the less room there is behind, but not how exactly. So I chose this way. Makes chance go down proportionally with number of viable tiles.